File: haskell/algebraic-polynomials/src/Math/Field/Transcendental.hs
-- |
-- Module      : Math.Field.Transcendental
-- Description : Transcendental extension fields Q(x1, ..., xn) via rational functions
-- Stability   : experimental
--
-- Elements of transcendental extension fields \(\mathbb{Q}(x_1, \ldots, x_n)\)
-- are represented as rational functions (ratios of multivariate polynomials).
--
-- When the coefficient type is 'Rational', fractions can be reduced to
-- lowest terms using 'reduceFrac' (which computes the multivariate GCD
-- of numerator and denominator).
--
-- Has 'Num', 'Fractional', 'Eq', and 'Ord' instances. The 'Ord' instance
-- provides a structural ordering for use in 'Map' keys and is /not/
-- mathematically meaningful.
module Math.Field.Transcendental
  ( RatFunc(..)
  , mkRatFunc
  , reduceFrac
  , constRF
  , varRF
  , isConstRF
  , evalRF
  ) where

import Data.Ratio (numerator, denominator)
import Math.Polynomial.Multivariate

-- | An element of \(\mathbb{Q}(x_1, \ldots, x_n)\): a ratio of
-- multivariate polynomials.
--
-- __Invariant:__ the denominator is non-zero.
--
-- For 'Rational' coefficients, use 'reduceFrac' to reduce to lowest terms.
data RatFunc k = RatFunc
  { rfNum :: !(MPoly k)  -- ^ Numerator polynomial.
  , rfDen :: !(MPoly k)  -- ^ Denominator polynomial (non-zero).
  } deriving (Show)

-- | Smart constructor ensuring a non-zero denominator.
--
-- Throws an error if the denominator is zero.
mkRatFunc :: MPoly k -> MPoly k -> RatFunc k
mkRatFunc _ d | isZero d = error "RatFunc: zero denominator"
mkRatFunc n d = RatFunc n d

-- | Reduce a rational function over 'Rational' to lowest terms by
-- dividing out the GCD of numerator and denominator.
--
-- Uses 'gcdMPoly' for the multivariate polynomial GCD.
reduceFrac :: RatFunc Rational -> RatFunc Rational
reduceFrac (RatFunc n d)
  | isZero n  = RatFunc zeroPoly (constPoly 1)
  | otherwise =
    let g = gcdMPoly n d
    in if isZero g || g == constPoly 1
       then RatFunc n d
       else RatFunc (exactDivMPoly n g) (exactDivMPoly d g)

-- | Construct a constant rational function from a scalar value.
constRF :: (Eq k, Num k) => k -> RatFunc k
constRF c = RatFunc (constPoly c) onePoly

-- | Construct a rational function representing a single transcendental
-- variable \(x_i\).
varRF :: (Eq k, Num k) => Var -> RatFunc k
varRF v = RatFunc (varPoly v) onePoly

-- | Test if a rational function is a constant (both numerator and
-- denominator are degree-0 polynomials with at most one term each).
isConstRF :: RatFunc k -> Bool
isConstRF (RatFunc n d) = numTerms n <= 1 && numTerms d == 1
    && totalDegree n == 0 && totalDegree d == 0

-- | Evaluate a rational function by substituting values for all variables.
--
-- The environment function maps each 'Var' to a value in @k@.
--
-- __Precondition:__ the denominator must not evaluate to zero.
evalRF :: (Eq k, Fractional k) => (Var -> k) -> RatFunc k -> k
evalRF env (RatFunc n d) = evalPoly env n / evalPoly env d

-- | Equality via cross-multiplication:
-- \(p_1/q_1 = p_2/q_2 \iff p_1 q_2 = p_2 q_1\).
instance (Eq k, Num k) => Eq (RatFunc k) where
  RatFunc n1 d1 == RatFunc n2 d2 = mulPoly n1 d2 == mulPoly n2 d1

-- | Structural ordering for use in 'Data.Map.Map' keys.
--
-- __Warning:__ this is /not/ a mathematically meaningful ordering.
instance (Eq k, Num k, Ord k) => Ord (RatFunc k) where
  compare (RatFunc n1 d1) (RatFunc n2 d2) =
    compare (unMPoly (mulPoly n1 d2)) (unMPoly (mulPoly n2 d1))

-- | Arithmetic on rational functions via 'Num'.
--
-- 'abs' and 'signum' are not meaningful and will throw errors.
instance (Eq k, Num k) => Num (RatFunc k) where
  -- p1/q1 + p2/q2 = (p1*q2 + p2*q1) / (q1*q2)
  RatFunc n1 d1 + RatFunc n2 d2 =
    RatFunc (addPoly (mulPoly n1 d2) (mulPoly n2 d1))
            (mulPoly d1 d2)

  -- p1/q1 * p2/q2 = (p1*p2) / (q1*q2)
  RatFunc n1 d1 * RatFunc n2 d2 =
    RatFunc (mulPoly n1 n2) (mulPoly d1 d2)

  negate (RatFunc n d) = RatFunc (negatePoly n) d

  abs    = error "RatFunc: abs not meaningful"
  signum = error "RatFunc: signum not meaningful"

  fromInteger = constRF . fromInteger

-- | 'Fractional' instance: 'recip' swaps numerator and denominator.
instance (Eq k, Num k) => Fractional (RatFunc k) where
  recip (RatFunc n d)
    | isZero n  = error "RatFunc: division by zero"
    | otherwise = RatFunc d n

  fromRational r =
    -- p/q as a RatFunc: numerator polynomial / denominator polynomial
    RatFunc (constPoly (fromInteger (numerator r)))
            (constPoly (fromInteger (denominator r)))

File: mercury/transcendental.m
%---------------------------------------------------------------------------%
% vim: ft=mercury ts=4 sw=4 et
%---------------------------------------------------------------------------%
% File: transcendental.m
% Main author: nyc
%
% Transcendental extension field Q(x1,...,xn).
%
% Elements are rational functions (ratios of multivariate polynomials).
% When the coefficient type is rational, fractions can be reduced to
% lowest terms using reduce_frac.
%
%---------------------------------------------------------------------------%

:- module transcendental.
:- interface.

:- import_module multivariate.
:- import_module poly.
:- import_module rational.

%---------------------------------------------------------------------------%
% Types
%---------------------------------------------------------------------------%

    % An element of Q(x1,...,xn): a ratio of multivariate polynomials.
    % Invariant: denominator is non-zero.
    %
:- type rat_func(K)
    --->    rat_func(
                rf_num :: mpoly(K),
                rf_den :: mpoly(K)
            ).

%---------------------------------------------------------------------------%
% Construction
%---------------------------------------------------------------------------%

    % Smart constructor: errors on zero denominator.
    %
:- func mk_rat_func(mpoly(K), mpoly(K)) = rat_func(K).

    % Constant rational function.
    %
:- func const_rf(K) = rat_func(K) <= ring(K).

    % A transcendental variable as a rational function.
    %
:- func var_rf(var) = rat_func(K) <= ring(K).

%---------------------------------------------------------------------------%
% Queries
%---------------------------------------------------------------------------%

    % Test if a rational function is constant.
    %
:- pred is_const_rf(rat_func(K)::in) is semidet.

%---------------------------------------------------------------------------%
% Evaluation
%---------------------------------------------------------------------------%

    % Evaluate a rational function by substituting values for variables.
    %
:- func eval_rf((func(var) = K), rat_func(K)) = K <= (ring(K), field(K)).

%---------------------------------------------------------------------------%
% Reduction
%---------------------------------------------------------------------------%

    % Reduce a rational function over rational to lowest terms.
    %
:- func reduce_frac(rat_func(rational)) = rat_func(rational).

%---------------------------------------------------------------------------%
% Arithmetic (ring/field instances)
%---------------------------------------------------------------------------%

:- instance ring(rat_func(K)) <= ring(K).
:- instance field(rat_func(K)) <= ring(K).

%---------------------------------------------------------------------------%
%---------------------------------------------------------------------------%

:- implementation.

:- use_module int.
:- import_module require.

%---------------------------------------------------------------------------%
% Construction
%---------------------------------------------------------------------------%

mk_rat_func(N, D) = Result :-
    ( if mp_is_zero(D) then
        unexpected($pred, "zero denominator in rat_func")
    else
        Result = rat_func(N, D)
    ).

const_rf(C) = rat_func(mp_const(C), mp_one).

var_rf(V) = rat_func(mp_var(V), mp_one).

%---------------------------------------------------------------------------%
% Queries
%---------------------------------------------------------------------------%

is_const_rf(rat_func(N, D)) :-
    int.'=<'(num_terms(N), 1),
    num_terms(D) = 1,
    total_degree(N) = 0,
    total_degree(D) = 0.

%---------------------------------------------------------------------------%
% Evaluation
%---------------------------------------------------------------------------%

eval_rf(Env, rat_func(N, D)) =
    field_div(mp_eval(Env, N), mp_eval(Env, D)).

%---------------------------------------------------------------------------%
% Reduction
%---------------------------------------------------------------------------%

reduce_frac(rat_func(N, D)) = Result :-
    ( if mp_is_zero(N) then
        Result = rat_func(mp_zero, mp_one)
    else
        G = gcd_mpoly(N, D),
        ( if mp_is_zero(G) then
            Result = rat_func(N, D)
        else if num_terms(G) = 1, total_degree(G) = 0 then
            Result = rat_func(N, D)
        else
            Result = rat_func(exact_div_mpoly(N, G), exact_div_mpoly(D, G))
        )
    ).

%---------------------------------------------------------------------------%
% Typeclass instances
%---------------------------------------------------------------------------%

:- instance ring(rat_func(K)) <= ring(K) where [
    (ring_zero = rat_func(mp_zero, mp_one)),
    (ring_one = rat_func(mp_one, mp_one)),
    % p1/q1 + p2/q2 = (p1*q2 + p2*q1) / (q1*q2)
    (ring_add(rat_func(N1, D1), rat_func(N2, D2)) =
        rat_func(mp_add(mp_mul(N1, D2), mp_mul(N2, D1)),
                 mp_mul(D1, D2))),
    % p1/q1 - p2/q2 = (p1*q2 - p2*q1) / (q1*q2)
    (ring_sub(rat_func(N1, D1), rat_func(N2, D2)) =
        rat_func(mp_sub(mp_mul(N1, D2), mp_mul(N2, D1)),
                 mp_mul(D1, D2))),
    % p1/q1 * p2/q2 = (p1*p2) / (q1*q2)
    (ring_mul(rat_func(N1, D1), rat_func(N2, D2)) =
        rat_func(mp_mul(N1, N2), mp_mul(D1, D2))),
    (ring_negate(rat_func(N, D)) = rat_func(mp_neg(N), D)),
    (ring_from_int(I) = rat_func(mp_const(ring_from_int(I)), mp_one)),
    (ring_is_zero(rat_func(N, _)) :- mp_is_zero(N))
].

:- instance field(rat_func(K)) <= ring(K) where [
    (field_div(A, rat_func(N2, D2)) = R :-
        ( if mp_is_zero(N2) then
            unexpected($pred, "division by zero in rat_func")
        else
            R = ring_mul(A, rat_func(D2, N2))
        ))
].

%---------------------------------------------------------------------------%
:- end_module transcendental.
%---------------------------------------------------------------------------%
